graph/path/bellman_ford_moore.go

```diff
@@ -27,46 +30,100 @@ func BellmanFordFrom(u graph.Node, g graph.Graph) (path Shortest, ok bool) {
 	path = newShortestFrom(u, nodes)
 	path.dist[path.indexOf[u.ID()]] = 0
 
+	// Queue to keep track which nodes need to be relaxed.
+	// Only nodes whose vertex distance changed in the previous iterations need to be relaxed again.
+	queue := newBellmanFordQueue(path.indexOf)
+	queue.enqueue(u)
+
+	n := len(nodes)
+	// The maximum of edges in a graph is |V| * (|V| -1) which is also the worst case complexity.
+	// If the queue-loop has more iterations than the amount of maximum edges
+	// it indicates that we have a negative cycle.
+	maxEdges := n * (n - 1)
+	negativeCycle := false
+
+	loops := 0
+
 	// TODO(kortschak): Consider adding further optimisations
 	// from http://arxiv.org/abs/1111.5414.
-	for i := 1; i < len(nodes); i++ {
-		changed := false
-		for j, u := range nodes {
-			uid := u.ID()
-			for _, v := range graph.NodesOf(g.From(uid)) {
-				vid := v.ID()
-				k := path.indexOf[vid]
-				w, ok := weight(uid, vid)
-				if !ok {
-					panic("bellman-ford: unexpected invalid weight")
-				}
-				joint := path.dist[j] + w
-				if joint < path.dist[k] {
-					path.set(k, joint, j)
-					changed = true
-				}
-			}
-		}
-		if !changed {
-			break
-		}
-	}
-
-	for j, u := range nodes {
+	for queue.len() != 0 {
+		u := queue.dequeue()
 		uid := u.ID()
+		j := path.indexOf[uid]
+
 		for _, v := range graph.NodesOf(g.From(uid)) {
 			vid := v.ID()
 			k := path.indexOf[vid]
 			w, ok := weight(uid, vid)
 			if !ok {
 				panic("bellman-ford: unexpected invalid weight")
 			}
-			if path.dist[j]+w < path.dist[k] {
-				path.hasNegativeCycle = true
-				return path, false
+
+			joint := path.dist[j] + w
+			if joint < path.dist[k] {
+				path.set(k, joint, j)
+
+				if !queue.has(vid) {
+					queue.enqueue(v)
+				}
 			}
 		}
+
+		if loops > maxEdges {
+			negativeCycle = true
+			break
+		}
+		loops++
+	}
+
+	if negativeCycle {
+		path.hasNegativeCycle = true
```

Comment:
This can be moved up to L73 rather than setting `negativeCycle = true` and breaking the loop, then `negativeCycle` can go away.

Reply:
Thanks, how could I not see that? I don't know...